### PR TITLE
sla: T4222: Add OWAMP and  TWAMP for service sla

### PR DESCRIPTION
**Pull request by sever-sever**
## Change Summary

OWAMP (https://tools.ietf.org/html/rfc4656 ) is a command-line client application and a policy daemon used to determine one-way latencies between hosts.
OWAMP session control uses traditional client-server communication between a control-client and a server,
TWAMP (two-way active measurement protocol) https://tools.ietf.org/html/rfc5357

Add configuration and operation modes

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes)
- [ ] Migration from an old Vyatta component to vyos-1x, please link to related PR inside obsoleted component
- [ ] Other (please describe):

## Related PR
Required additional pkgs (task in vyos-build)
 owamp-client_4.4.4-1_amd64.deb
 owamp-server_4.4.4-1_amd64.deb
 twamp-client_4.4.4-1_amd64.deb
 twamp-server_4.4.4-1_amd64.deb

https://github.com/vyos/vyos-build/pull/239

## Related Task(s)

* https://phabricator.vyos.net/T4222

## Component(s) name

owamp, twamp
## How to test
VyOS sever configuration, set one or both sla servers
```
set service sla owamp-server
set service sla twamp-server
```
From the client site ping a server
```
force owping 192.0.2.120
force twping 192.0.2.190
```

Check services/pids:
```
vyos@r14# ps ax | grep amp
   5049 ?        Ss     0:00 /usr/sbin/owampd -c /etc/owamp-server -R /var/run
   5070 ?        Ss     0:00 /usr/sbin/twampd -c /etc/twamp-server -R /var/run
```

```
vyos@r14:~$ force twping 192.168.122.12
Approximately 12.9 seconds until results available

Directional delays may be inaccurate due to out of sync clocks!

--- twping statistics from [192.168.122.190]:9718 to [192.168.122.12]:9718 ---
SID:	c0a87a0ce64217c616e87d2c0e0e1053
first:	2022-06-01T19:52:23.321
last:	2022-06-01T19:52:33.788
100 sent, 0 lost (0.000%), 0 send duplicates, 0 reflect duplicates
round-trip time min/median/max = 0.446/0.7/0.86 ms, (err=1.38 ms)
send time min/median/max = -294/-293/-293 ms, (err=0.69 ms)
reflect time min/median/max = 294/294/294 ms, (err=0.69 ms)
reflector processing time min/max = 0.00238/0.00811 ms
two-way jitter = 0.2 ms (P95-P50)
send jitter = 0.1 ms (P95-P50)
reflect jitter = 0.1 ms (P95-P50)
send hops = 0 (consistently)
reflect hops = 0 (consistently)
```

## Checklist:

- [x] I have read the [**CONTRIBUTING**](https://github.com/vyos/vyos-1x/blob/current/CONTRIBUTING.md) document
- [x] I have linked this PR to one or more Phabricator Task(s)
- [ ] I have run the components [**SMOKETESTS**](https://github.com/vyos/vyos-1x/tree/current/smoketest/scripts/cli) if applicable
- [x] My commit headlines contain a valid Task id
- [ ] My change requires a change to the documentation
- [ ] I have updated the documentation accordingly

### CODE REVIEW

File: src/conf_mode/service_sla.py [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+from vyos.configdict import dict_merge
+from vyos.template import render
+from vyos.util import call
+from vyos.xml import defaults
+from vyos import ConfigError
+from vyos import airbag
+airbag.enable()
+
+
+owamp_config_dir = '/etc/owamp-server'
```
Comment (c-po) [inline src/conf_mode/service_sla.py:31]: Please render configs to `/etc/owamp-server`

File: src/conf_mode/service_sla.py [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+    render(systemd_override_owamp, 'sla/owamp-override.conf.j2', sla)
+
+    render(twamp_config_file, 'sla/twamp-server.conf.j2', sla)
+    render(systemd_override_twamp, 'sla/twamp-override.conf.j2', sla)
+
+    return None
+
+def apply(sla):
+    if not sla:
+        call('systemctl daemon-reload')
```
Comment (c-po) [inline src/conf_mode/service_sla.py]: You could actually make the if/else nesting a lot easier, and remove duplicates for e.g. removing stale config files.


```python
def apply(sla):
    owamp_service = 'owamp-server.service'
    twamp_service = 'twamp-server.service'
    
    call('systemctl daemon-reload')

    if not sla or 'owamp_server' not in sla:
        call(f'systemctl stop {owamp_service}')

        if os.path.exists(owamp_config_file):
            os.unlink(owamp_config_file)
    
    if not sla or 'twamp_server' not in sla:    
        call(f'systemctl stop {twamp_service}')            
        if os.path.exists(twamp_config_file):
            os.unlink(twamp_config_file)

    if 'owamp_server' in sla:
        call(f'systemctl reload-or-restart {owamp_service}')

    if 'twamp_server' in sla:
        call(f'systemctl reload-or-restart {twamp_service}')

    return None
```

**Review by c-po — CHANGES_REQUESTED** [review]

File: src/conf_mode/service_sla.py (hunk 1 above)
Comment (sever-sever) [reply src/conf_mode/service_sla.py:31]: To start each daemon we set parameter configDir
`ExecStart=/usr/sbin/owampd -c /etc/owamp-server -R /var/run`
And not config file itself.
So I think dirs should be different as each daemon reads own files in config directory

File: src/conf_mode/service_sla.py (hunk 2 above)
Comment (sever-sever) [reply src/conf_mode/service_sla.py]: Issue with delete sla
`delete service sla`
```
Traceback (most recent call last):
  File "/usr/libexec/vyos/conf_mode/service_sla.py", line 111, in <module>
    apply(c)
  File "/usr/libexec/vyos/conf_mode/service_sla.py", line 97, in apply
    if 'owamp_server' in sla:
TypeError: argument of type 'NoneType' is not iterable

```
update, fixed

Comment (c-po) [reply src/conf_mode/service_sla.py]: Use

```python
    if sla and  'owamp_server' in sla:
        call(f'systemctl reload-or-restart {owamp_service}')

    if sla and 'twamp_server' in sla:
        call(f'systemctl reload-or-restart {twamp_service}')
```

instead

**Pull request merged**